### ISSUE TITLE
[R] Lambdarank cross validation

### ISSUE THREAD

**Issue by ntarasov** [Microsoft/LightGBM#390]
Hi,
when I use lgb.cv for cross validation lambdarank objective:
```
train_data <- lgb.Dataset(train_matrix,
                          label = train$relevance,
                          group = groups$count,
                          colnames = colnames(train_matrix),
                          free_raw_data = F)
params <- list(objective = 'lambdarank',
               boosting_type = 'gbdt',
               label_gain = '0,1,2,3,4,5,6,7,8,9,10',
               metric = "ndcg",
               metric_freq = 1,
               is_training_metric = T,
               ndcg_at = '10',
               learning_rate = 0.1,
               verbosity = 10)
model <- lgb.cv(params, train_data, nrounds = 2, folds = data_folds)
```
I get the result:
```
[LightGBM] [Info] Total Bins 18010
[LightGBM] [Info] Number of data: 53105, number of used features: 147
[LightGBM] [Info] Total Bins 18010
[LightGBM] [Info] Number of data: 47848, number of used features: 147
[LightGBM] [Info] Total Bins 18010
[LightGBM] [Info] Number of data: 52811, number of used features: 147
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=1 and max_depth=1
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements.
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=1 and max_depth=1
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements.
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=1 and max_depth=1
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements.
[1]:	valid's ndcg@10:NaN+NaN 
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=1 and max_depth=1
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements.
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=1 and max_depth=1
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements.
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=1 and max_depth=1
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements.
[2]:	valid's ndcg@10:NaN+NaN 
```

But, when I use train with validation:
`model <- lgb.train(params, train_data, valids = list(test = train_data), nrounds = 400)`
I get normal result:
```
[LightGBM] [Info] Total Bins 18010
[LightGBM] [Info] Number of data: 76882, number of used features: 147
[LightGBM] [Info] Trained a tree with leaves=127 and max_depth=17
[1]:	test's ndcg@10:0.738216 
[LightGBM] [Info] Trained a tree with leaves=127 and max_depth=14
[2]:	test's ndcg@10:0.808526 
```
Dataset labels: 0,1,2,3,4,9,10

So what's the reason?

**Comment by ntarasov** [comment]
Package lightgbm version 0.1

**Comment by guolinke** [comment]
what is your label range? refer to #387 .

**Comment by ntarasov** [comment]
Dataset labels: 0,1,2,3,4,9,10

**Comment by guolinke** [comment]
@Laurae2 
It seems there are some bugs on CV. Can you help to check that?

**Comment by guolinke** [comment]
I take a look. Both python/R have bugs for the Cross validation for ranking task.
The reason is it makes folds randomly without considering group information.
The correct approach is making folds by groups, not by data.

**Comment by ntarasov** [comment]
@guolinke I use custom data split considering group information, because nfold causes message:
```
Error in generate.cv.folds(nfold, nrow(data), stratified, getinfo(data,  : 
  
	Automatic generation of CV-folds is not implemented for lambdarank!
	Consider providing pre-computed CV-folds through the 'folds=' parameter.
```

This is data split:
> str(data_folds)
List of 3
 $ : int [1:23777] 89 90 91 92 93 94 95 96 97 98 ...
 $ : int [1:29034] 1 2 3 4 5 6 7 8 9 10 ...
 $ : int [1:24071] 448 449 450 451 452 453 454 455 456 457 ...
> length(unlist(data_folds))
[1] 76882
> length(unique(unlist(data_folds)))
[1] 76882

**Comment by Laurae2** [comment]
@guolinke xgboost doesn't implement CV in ranking tasks with query/group because it requires reconstruction of the appropriate datasets each time (this is not a major problem), along with all the issues it brings (this is a major problem). For instance, if you pass a group information, xgboost will always refuse to perform CV.

However, I'm trying to go through, check #397. Trying to find out if there is a solution to it (other than performing CV fold by fold, like what xgboost requires), but I'm not seeing it any currently.

**Comment by guolinke** [comment]
@wxchan 
can you help to fix this in python package ? we should sample based on group_id. The data in the same group should all be sampled, or all not be sampled.

**Comment by wxchan** [comment]
@guolinke do you mean something like http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GroupKFold.html

**Comment by guolinke** [comment]
@wxchan yea, exactly.

**Comment by guolinke** [comment]
@Laurae2 any updates for this ?
